tests/ui_tools/test_boxes.py

```diff
@@ -99,28 +99,54 @@ def test_generic_autocomplete_mentions(self, write_box, text,
         typeahead_string = write_box.generic_autocomplete(text, state)
         assert typeahead_string == required_typeahead
 
-    @pytest.mark.parametrize('text, state, required_typeahead', [
-        ('#Stream', 0, '#**Stream 1**'),
-        ('#Stream', 1, '#**Stream 2**'),
-        ('#S', 0, '#**Some general stream**'),
-        ('#S', 1, '#**Secret stream**'),
-        ('#S', 2, '#**Stream 1**'),
-        ('#S', 3, '#**Stream 2**'),
-        ('#S', -1, '#**Stream 2**'),
-        ('#S', -2, '#**Stream 1**'),
-        ('#S', -3, '#**Secret stream**'),
-        ('#S', -4, '#**Some general stream**'),
-        ('#S', -5, None),
-        ('#So', 0, '#**Some general stream**'),
-        ('#So', 1, None),
-        ('#Se', 0, '#**Secret stream**'),
-        ('#Se', 1, None),
-        ('#St', 0, '#**Stream 1**'),
-        ('#St', 1, '#**Stream 2**'),
-        ('#Stream 1', 0, '#**Stream 1**'),
+    @pytest.mark.parametrize('text, state, required_typeahead, to_pin', [
+        # With no streams pinned.
+        ('#Stream', 0, '#**Stream 1**', []),  # 1st-word startswith match.
+        ('#Stream', 1, '#**Stream 2**', []),  # 1st-word startswith match.
+        ('#Stream', 2, '#**Secret stream**', []),  # 2nd-word startswith match.
+        ('#Stream', 3, '#**Some general stream**', []),  # 3rd-word startswith.
+        ('#S', 0, '#**Some general stream**', []),  # 1st-word startswith.
+        ('#S', 1, '#**Secret stream**', []),  # 1st-word startswith match.
```

Comment:
Should the above two be reversed? ('Se' before 'So')

So, an equal match is sorted. This is relatively minor, but it's unclear how it's sorting otherwise?

Reply:
To clarify, `match_stream` doesn't sort; it expects `streams` to be sorted (I have mentioned this in the docstring).

Re the 'Se' before 'So', 'Some general stream' has been put before 'Secret stream' in `streams_fixture` which is why the suggestions are being given in this order.

```python
@pytest.fixture
def streams_fixture():
    streams = [general_stream, secret_stream]
    ...
```

Reply:
I agree. We could either change the name from 'Some general stream' to 'General Stream' (that'd fix the order) or sort the streams in the fixture itself. 

The assert sounds appealing. It would require the input to be sorted while maintaining the pinning order (e.g. `assert input = sorted(input, key=is_pinned)`), right?. In that case, would the assert be better than actually sorting the input?

Reply:
The fixture is supposed to represent what is present in the running code, so it should probably be in the right order in any case, given that's what the function expects. I think that's probably the minimum, whether we adjust the fixture or potentially change the name - though I don't know if the latter approach would affect the scope of other tests that use the same fixture.

The assert would be a code-based representation of what the range of inputs are supposed to be, a little like the +1/-1 in set_counts. Sometimes that can be expressed more cleanly in terms of the data type used, but failing that then it's useful to ensure that eg. in that case we don't pass +5 to set _count!